src/tools/illink/src/linker/Linker/TypeNameResolver.cs

```diff
@@ -5,22 +5,33 @@
 using System.Collections.Generic;
 using System.Diagnostics;
 using System.Diagnostics.CodeAnalysis;
-using System.Reflection.Runtime.TypeParsing;
 using ILLink.Shared;
 using ILLink.Shared.TrimAnalysis;
 using Mono.Cecil;
+using Mono.Linker;
+
+using TypeName = System.Reflection.Metadata.TypeName;
+using TypeNameParseOptions = System.Reflection.Metadata.TypeNameParseOptions;
+using AssemblyNameInfo = System.Reflection.Metadata.AssemblyNameInfo;
+using TypeNameHelpers = System.Reflection.Metadata.TypeNameHelpers;
+
+#nullable enable
 
 namespace Mono.Linker
 {
 	internal sealed class TypeNameResolver
 	{
-		readonly LinkContext _context;
+		readonly ITryResolveMetadata _metadataResolver;
+		readonly ITryResolveAssemblyName _assemblyResolver;
+
+		private static readonly TypeNameParseOptions s_typeNameParseOptions = new () { MaxNodes = int.MaxValue };
```

Comment:
cc @adamsitnik @GrabYourPitchforks So far every single real-world use of `TypeName.Parse` except the one in NRBF overrides the low MaxNodes default with int.MaxValue. It makes me wonder whether we made the right choice for the default.

Reply:
`int.MaxValue` is the right value to use for tools like ILLinker. The tools like ILLinker are not hardened against too complex input. It is fine for these tools to crash with StackOverflowException or take a long time to run when the input is too complex.

Reply:
> So far every single real-world use of TypeName.Parse except the one in NRBF overrides the low MaxNodes default with int.MaxValue. It makes me wonder whether we made the right choice for the default.

From my perspective, we wanted to be secure by default and this was the right choice.

IMO `int.MaxValue` is never the right value, as it is impossible to create such complex type names (runtime is unable to load them). For Runtime and its tools we should rather find a value that is large enough to allow for the most complex real-life types, but not big enough to cause SO.

Reply:
> For Runtime and its tools we should rather find a value that is large enough to allow for the most complex real-life types, but not big enough to cause SO.

There is no point in doing that unless we make the runtime and tools hardened against too complex types. There are 1000 different ways how too complex types can introduce stackoverflow in the runtime and tools. Reducing that by one does not really help.

Reply:
> Reducing that by one does not really help.

It does not solve the problem, but on the other hand it's almost free (the logic is implemented, it's just a matter of changing one value in few places) and it would not hurt.

But, I don't have the perspective of long term thinking about the .NET Runtime you have, so I am most probably missing a lot of details (like giving the users a false impression that `Type.GetType(string)` is hardened against all kinds of unsafe inputs)